libpod/container_exec.go

```diff
@@ -775,12 +776,30 @@ func (c *Container) Exec(config *ExecConfig, streams *define.AttachStreams, resi
 }
 
 // cleanup an exec session after its done
-func (c *Container) cleanupExecBundle(sessionID string) error {
-	if err := os.RemoveAll(c.execBundlePath(sessionID)); err != nil && !os.IsNotExist(err) {
-		return err
+func (c *Container) cleanupExecBundle(sessionID string) (Err error) {
+	path := c.execBundlePath(sessionID)
+	for attempts := 0; attempts < 50; attempts++ {
+		Err = os.RemoveAll(path)
+		if Err == nil || os.IsNotExist(Err) {
+			return nil
+		}
+		if pathErr, ok := Err.(*os.PathError); ok {
+			Err = pathErr.Err
+			if errors.Cause(Err) == unix.ENOTEMPTY || errors.Cause(Err) == unix.EBUSY {
+				// give other processes a chance to use the container
+				if !c.batched {
+					c.lock.Unlock()
```

Comment:
I don't know if this is safe, we open the door to having the container modified while we're running. Any code-path that calls this could suddenly find itself trying to interact with a container in a very different state than what we left; leaves the door open to DB corruption, as we don't actually do any validation checks on a lot of our writes to the DB as we assume we hold the lock for the complete duration of the operation's lifetime.

If you do this, you definitely need a `c.syncContainer()` after you re-lock the container to ensure the state remains synced.

Reply:
I've added a `c.syncContainer()`